The tar command fails on a Windows runner

**Describe the bug**
The tar command fails on a Windows runner.

**To Reproduce**
Steps to reproduce the behavior:
1. Go to https://github.com/exivity/rabbitmq-windows-docker/runs/966529913
2. Click on _Post Run satackey/action-docker-layer-caching@v0.0.5_
3. Expand the step starting with `sh -c docker save [...]`
4. See error

**Expected behavior**
The Post Run to succeed

**Debug logs**
```
sh -c docker save '1ea7049bf5c6' '1ea7049bf5c6' '598c98f7b656' '912539758c76' 'dcbc260c6b09' 'a3cb6ed51c1b' 'e3de158569a9' '399de6e28bb0' '8809fb27bbb2' '4df9dd644716' '148c3b76ab54' '2b250bc046c1' 'e5e57296b9ca' '1ff1fd5e1fa2' '866b450c8b64' '68d6ccbac8d9' '7ca015404fc3' '318b2224022d' '5b43130792e4' '6c12dc0a117c' '987b1d5e0abf' 'exivity/rabbitmq:3.8.6' '1ea7049bf5c6' '598c98f7b656' '912539758c76' 'dcbc260c6b09' 'a3cb6ed51c1b' 'e3de158569a9' '399de6e28bb0' '8809fb27bbb2' '4df9dd644716' '148c3b76ab54' '2b250bc046c1' 'e5e57296b9ca' '1ff1fd5e1fa2' '866b450c8b64' '68d6ccbac8d9' '7ca015404fc3' '318b2224022d' '5b43130792e4' '6c12dc0a117c' '987b1d5e0abf' | tar xf - -C d:\a\rabbitmq-windows-docker\rabbitmq-windows-docker\.action-docker-layer-caching-docker_images\image
  "C:\Program Files\Git\bin\sh.exe" -c "docker save '1ea7049bf5c6' '1ea7049bf5c6' '598c98f7b656' '912539758c76' 'dcbc260c6b09' 'a3cb6ed51c1b' 'e3de158569a9' '399de6e28bb0' '8809fb27bbb2' '4df9dd644716' '148c3b76ab54' '2b250bc046c1' 'e5e57296b9ca' '1ff1fd5e1fa2' '866b450c8b64' '68d6ccbac8d9' '7ca015404fc3' '318b2224022d' '5b43130792e4' '6c12dc0a117c' '987b1d5e0abf' 'exivity/rabbitmq:3.8.6' '1ea7049bf5c6' '598c98f7b656' '912539758c76' 'dcbc260c6b09' 'a3cb6ed51c1b' 'e3de158569a9' '399de6e28bb0' '8809fb27bbb2' '4df9dd644716' '148c3b76ab54' '2b250bc046c1' 'e5e57296b9ca' '1ff1fd5e1fa2' '866b450c8b64' '68d6ccbac8d9' '7ca015404fc3' '318b2224022d' '5b43130792e4' '6c12dc0a117c' '987b1d5e0abf' | tar xf - -C d:\a\rabbitmq-windows-docker\rabbitmq-windows-docker\.action-docker-layer-caching-docker_images\image"
  tar: d\:arabbitmq-windows-dockerrabbitmq-windows-docker.action-docker-layer-caching-docker_imagesimage: Cannot open: No such file or directory
  tar: Error is not recoverable: exiting now
  write /dev/stdout: The pipe is being closed.
  Error: The process 'C:\Program Files\Git\bin\sh.exe' failed with exit code 2
      at ExecState._setResult (d:\a\_actions\satackey\action-docker-layer-caching\v0.0.5\dist\ttsc-dist\post.js\index.js:1300:25)
      at ExecState.CheckComplete (d:\a\_actions\satackey\action-docker-layer-caching\v0.0.5\dist\ttsc-dist\post.js\index.js:1283:18)
      at ChildProcess.<anonymous> (d:\a\_actions\satackey\action-docker-layer-caching\v0.0.5\dist\ttsc-dist\post.js\index.js:1183:27)
      at ChildProcess.emit (events.js:210:5)
      at maybeClose (internal/child_process.js:1021:16)
      at Socket.<anonymous> (internal/child_process.js:430:11)
      at Socket.emit (events.js:210:5)
      at Pipe.<anonymous> (net.js:659:12)
  ##[error]Error: The process 'C:\Program Files\Git\bin\sh.exe' failed with exit code 2
```

**Runner Environment (please complete the following information):**
 - OS: `windows-2019`
 - Action version: `v0.0.5`

**Additional context**
Seems that the arguments to the tar command are not properly escaped on Windows which causes all directories to be concatenated (where path separators are expected). I think the relevant is here: https://github.com/satackey/action-docker-layer-caching/blob/master/src/LayerCache.ts#L59

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -56,9 +56,6 @@ jobs:
               'ubuntu-latest',
               'windows-latest',
             ],
-            branch: [
-              context.payload.ref.replace('refs/heads/', '')
-            ],
             include: [
               {
                 inspect_image: 'test_project_scratch',
@@ -68,7 +65,9 @@ jobs:
                 inspect_image: 'hello-world',
                 prepare_command: ':',
                 build_command: 'docker pull hello-world',
-              },
+              }, {
+                branch: process.env.GITHUB_REF.replace('refs/heads/', '')
+              }
             ],
             exclude: [
               {

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -48,6 +48,13 @@ jobs:
               'test_project_scratch',
               'hello-world',
             ],
+            os: [
+              'ubuntu-latest',
+              'windows-latest',
+            ],
+            branch: [
+              context.payload.ref.replace('refs/heads/', '')
+            ],
             include: [
               {
                 inspect_image: 'test_project_scratch',
@@ -69,18 +76,13 @@ jobs:
     needs: build
     strategy:
       matrix: ${{ fromJSON(needs.build.outputs.matrix) }}
-    runs-on: ubuntu-latest
+    runs-on: ${{ matrix.os }}
 
     steps:
     - uses: actions/checkout@v2
 
     - run: ${{ matrix.prepare_command }}
 
-    - name: Extract
-      id: extract
-      run: |
-        echo "##[set-output name=branch;]${GITHUB_REF#refs/heads/}"
-    
     - name: Download action
       uses: actions/download-artifact@v2
       with:
@@ -88,7 +90,7 @@ jobs:
         path: action-dlc
 
     - uses: ./action-dlc
-      name: Run satackey/action-docker-layer-caching@${{ steps.extract.outputs.branch }}
+      name: Run satackey/action-docker-layer-caching@${{ matrix.branch }}
       with:
         key: docker-layer-caching-${{ matrix.inspect_image }}-sha:${{ github.sha }}-{hash}
 
@@ -98,7 +100,7 @@ jobs:
     needs: [build, test_saving]
     strategy:
       matrix: ${{ fromJSON(needs.build.outputs.matrix) }}
-    runs-on: ubuntu-latest
+    runs-on: ${{ matrix.os }}
     steps:
     - uses: actions/checkout@v2
 
@@ -116,7 +118,7 @@ jobs:
         path: action-dlc
 
     - uses: ./action-dlc
-      name: Run satackey/action-docker-layer-caching@${{ steps.extract.outputs.branch }}
+      name: Run satackey/action-docker-layer-caching@${{ matrix.branch }}
       with:
         key: never-restored-docker-layer-caching-${{ matrix.inspect_image }}-sha:${{ github.sha }}-{hash}
         restore-keys: docker-layer-caching-${{ matrix.inspect_image }}-sha:${{ github.sha }}-

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -70,6 +70,12 @@ jobs:
                 build_command: 'docker pull hello-world',
               },
             ],
+            exclude: [
+              {
+                inspect_image: 'test_project_scratch',
+                os: 'windows-latest',
+              },
+            ],
           }
 
     outputs:

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -111,6 +111,7 @@ jobs:
     strategy:
       matrix: ${{ fromJSON(needs.build.outputs.matrix) }}
     runs-on: ${{ matrix.os }}
+
     steps:
     - uses: actions/checkout@v2
 

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -7,6 +7,10 @@ on:
   pull_request:
   delete:
 
+defaults:
+  run:
+    shell: bash
+
 jobs:
   build:
     runs-on: ubuntu-latest

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -102,7 +102,7 @@ jobs:
     - uses: ./action-dlc
       name: Run satackey/action-docker-layer-caching@${{ matrix.branch }}
       with:
-        key: docker-layer-caching-${{ matrix.inspect_image }}-sha:${{ github.sha }}-{hash}
+        key: docker-layer-caching-${{ matrix.os }}-${{ matrix.inspect_image }}-sha:${{ github.sha }}-{hash}
 
     - run: ${{ matrix.build_command }}
 
@@ -130,8 +130,8 @@ jobs:
     - uses: ./action-dlc
       name: Run satackey/action-docker-layer-caching@${{ matrix.branch }}
       with:
-        key: never-restored-docker-layer-caching-${{ matrix.inspect_image }}-sha:${{ github.sha }}-{hash}
-        restore-keys: docker-layer-caching-${{ matrix.inspect_image }}-sha:${{ github.sha }}-
+        key: never-restored-docker-layer-caching-${{ matrix.os }}-${{ matrix.inspect_image }}-sha:${{ github.sha }}-{hash}
+        restore-keys: docker-layer-caching-${{ matrix.os }}-${{ matrix.inspect_image }}-sha:${{ github.sha }}-
         skip-save: 'true'
 
     - name: Show cached image info

--- a/src/ImageDetector.ts
+++ b/src/ImageDetector.ts
@@ -5,7 +5,7 @@ export class ImageDetector {
   async getExistingImages(): Promise<string[]> {
     const existingSet = new Set<string>([])
     const ids = (await exec.exec(`docker image ls -q`, [], { silent: true })).stdoutStr.split(`\n`).filter(id => id !== ``)
-    const repotags = (await exec.exec(`sh -c "docker image ls --format '{{ .Repository }}:{{ .Tag }}' --filter 'dangling=false'"`, [], { silent: true })).stdoutStr.split(`\n`).filter(id => id !== ``);
+    const repotags = (await exec.exec(`docker`, `image ls --format {{.Repository}}:{{.Tag}} --filter dangling=false`.split(' '), { silent: true })).stdoutStr.split(`\n`).filter(id => id !== ``);
     core.debug(JSON.stringify({ log: "getExistingImages", ids, repotags }));
     ([...ids, ...repotags]).forEach(image => existingSet.add(image))
     core.debug(JSON.stringify({ existingSet }))

--- a/src/LayerCache.ts
+++ b/src/LayerCache.ts
@@ -133,14 +133,15 @@ class LayerCache {
     }
   }
 
-  private async storeSingleLayerBy(id: string): Promise<number> {
-    const path = this.genSingleLayerStorePath(id)
-    const key = await this.generateSingleLayerSaveKey(id)
+  private async storeSingleLayerBy(layerId: string): Promise<number> {
+    const path = this.genSingleLayerStorePath(layerId)
+    const key = await this.generateSingleLayerSaveKey(layerId)
 
-    core.info(`Start storing layer cache: ${key}`)
+    core.info(`Start storing layer cache: ${JSON.stringify({ layerId, key })}`)
     const cacheId = await LayerCache.dismissError(cache.saveCache([path], key), LayerCache.ERROR_CACHE_ALREAD_EXISTS_STR, -1)
-    core.info(`Stored layer cache, key: ${key}, id: ${cacheId}`)
+    core.info(`Stored layer cache: ${JSON.stringify({ key, cacheId })}`)
 
+    core.debug(JSON.stringify({ log: `storeSingleLayerBy`, layerId, path, key, cacheId}))
     return cacheId
   }
 
@@ -201,9 +202,12 @@ class LayerCache {
   }
 
   private async restoreSingleLayerBy(id: string): Promise<string> {
-    core.debug(JSON.stringify({ log: `restoreSingleLayerBy`, id }))
+    const path = this.genSingleLayerStorePath(id)
+    const key = await this.recoverSingleLayerKey(id)
 
-    const result = await cache.restoreCache([this.genSingleLayerStorePath(id)], await this.recoverSingleLayerKey(id))
+    core.debug(JSON.stringify({ log: `restoreSingleLayerBy`, id, path, key }))
+
+    const result = await cache.restoreCache([path], key)
 
     if (result == null) {
       throw new Error(`${LayerCache.ERROR_LAYER_CACHE_NOT_FOUND_STR}: ${JSON.stringify({ id })}`)

--- a/src/LayerCache.ts
+++ b/src/LayerCache.ts
@@ -178,8 +178,9 @@ class LayerCache {
   }
 
   private async restoreLayers(): Promise<boolean> {
-    const pool = new PromisePool(this.concurrency)
 
+    
+    const pool = new PromisePool(this.concurrency)
     const tasks = (await this.getLayerIds()).map(
       layerId => pool.open(() => this.restoreSingleLayerBy(layerId))
     )
@@ -204,9 +205,11 @@ class LayerCache {
   private async restoreSingleLayerBy(id: string): Promise<string> {
     const path = this.genSingleLayerStorePath(id)
     const key = await this.recoverSingleLayerKey(id)
+    const dir = path.replace(/[^/\\]+$/, ``)
 
-    core.debug(JSON.stringify({ log: `restoreSingleLayerBy`, id, path, key }))
+    core.debug(JSON.stringify({ log: `restoreSingleLayerBy`, id, path, dir, key }))
 
+    await fs.mkdir(dir, { recursive: true })
     const result = await cache.restoreCache([path], key)
 
     if (result == null) {
@@ -247,7 +250,7 @@ class LayerCache {
   }
 
   genSingleLayerStorePath(id: string) {
-    return `${this.getLayerCachesDir()}/${id}/layer.tar`
+    return path.resolve(`${this.getLayerCachesDir()}/${id}/layer.tar`)
   }
 
   async generateRootHashFromManifest(): Promise<string> {

--- a/src/LayerCache.ts
+++ b/src/LayerCache.ts
@@ -49,8 +49,8 @@ class LayerCache {
   }
 
   private async saveImageAsUnpacked() {
-    await this.exec('mkdir -p', [this.getSavedImageTarDir()], { silent: true })
-    await this.exec(`sh -c`, [`docker save '${(await this.makeRepotagsDockerSaveArgReady(this.ids)).join(`' '`)}' | tar xf - -C ${this.getSavedImageTarDir()}`])
+    await fs.mkdir(this.getSavedImageTarDir(), { recursive: true })
+    await this.exec(`sh -c`, [`docker save '${(await this.makeRepotagsDockerSaveArgReady(this.ids)).join(`' '`)}' | tar xf - -C .`], { cwd: this.getSavedImageTarDir() })
   }
 
   private async makeRepotagsDockerSaveArgReady(repotags: string[]): Promise<string[]> {

--- a/src/Tar.ts
+++ b/src/Tar.ts
@@ -16,6 +16,7 @@ export function assertManifests(x: unknown): asserts x is Manifests {
 export async function loadRawManifests(path: string) {
   return (await fs.readFile(`${path}/manifest.json`)).toString()
 }
+
 export async function loadManifests(path: string) {
   const raw = await loadRawManifests(path)
   const manifests = JSON.parse(raw.toString())

